### ISSUE TITLE
Seek preview issues

### ISSUE THREAD

**Issue by clsid2** [clsid2/mpc-hc#835]
@adipose

When moving mouse while preview is visible, and preview video isn't decoded fast enough, then the black background of the preview window is visible in between the video stills. So it sort of flickers. Could we avoid redrawing the background? Just keep the old video image until new one is ready.

Middle mouse buttons allow disabling the preview. Which is fine. But I noticed it actually disables the whole option. Maybe it should only hide the window. Then it would also be possible to toggle it on again.

I saw a report where the preview window shows, but says black, even when holding in a static position. Reloading same file makes it work again, so it is sort of random. I have not been able to reproduce. Any idea what might be wrong?

**Comment by adipose** [comment]
> When moving mouse while preview is visible, and preview video isn't decoded fast enough, then the black background of the preview window is visible in between the video stills. So it sort of flickers. Could we avoid redrawing the background? Just keep the old video image until new one is ready.

The paint method doesn't actually draw the background.  At one time I had an erase method to make it redraw because it left a "hollow" background and you could see the main video behind it.  But that turned out to be unnecessary.  The paint method is supposed to exclude the video rect, but maybe the flicker occurs because the video rect is wrong somehow.  I can create a test build to see.


> 
> Middle mouse buttons allow disabling the preview. Which is fine. But I noticed it actually disables the whole option. Maybe it should only hide the window. Then it would also be possible to toggle it on again.

This was some odd code from -be, I left it in for now but I don't know if I like it, as it could conflict with actual shortcuts set by the user.  My thought was to create a key to toggle it, then let them assign what they want.  But it should be improved probably.

> 
> I saw a report where the preview window shows, but says black, even when holding in a static position. Reloading same file makes it work again, so it is sort of random. I have not been able to reproduce. Any idea what might be wrong?

Maybe the member variable for the video size didn't get updated.

**Comment by adipose** [comment]
I just tested in debug mode on a DVD, which is a bit slower to get the frames, and I observed the transparent background issue, but no black background.

**Comment by adipose** [comment]
@clsid2 , try this:

https://github.com/adipose/mpc-hc/commits/patch173

I made sure that the default erase bg cannot be called.  See if that solves the flicker issue for you.  In testing, it does seem *marginally* better to me, but I'm not sure.

**Comment by clsid2** [comment]
I didn't have the flicker myself, but I will ask someone to test it.

Another issue that I noticed is that when the preview is displayed for the first time, I see a resizing glitch of the video inside the preview. Doesn't happen again after reloading a file. Only after restarting the player.

**Comment by ale5000-git** [comment]
Do you have a test build?

**Comment by adipose** [comment]
https://mega.nz/file/YUpy0byA#Pfbcg2M2Oai4LFWqjtAlk3dbfMdezVAVtdERjO1gVOg

**Comment by ale5000-git** [comment]
Unfortunately the flickers remains and in addition I also managed to get the preview window trasparent: 
![transparent-preview-window](https://user-images.githubusercontent.com/15793015/104394511-0fbcc300-5547-11eb-8674-2102c7caef9b.png)

**Comment by MadokaAyukawa** [comment]
I also get the flicker with the test build.

I accidentally found the middle mouse click turning off the preview option as I assign a middle mouse click to a function. The BE feature doesn't allow one to turn the preview back on with a second middle mouse click, so it's annoying that it's one way only.

**Comment by RedCard1nal** [comment]
I used to use MPC-BE and experienced the same flickering problem. I haven't done too much testing so far but the amount of flickering seems to vary depending on the video, which I don't remember being the case in MPC-BE. Whenever I play some old Star Trek DVD x264 encodes there is basically no flickering...

**Comment by clsid2** [comment]
Which GPUs do you all have? Which windows version?

**Comment by RedCard1nal** [comment]
> 
> 
> Which GPUs do you all have? Which windows version?

Windows 10 20H2 (19042), GTX 1060 6GB

**Comment by MadokaAyukawa** [comment]
Win 10 1909 with Intel UHD Graphics 620.

I have an old computer which does not seem to flicker:
Win 7 with an ATI Radeon HD 4850

**Comment by clsid2** [comment]
In case of NVIDIA enable "Vertical Sync" in the NVIDIA driver settings. That can help with stuttering issues. Maybe it also makes a difference for this flickering issue.

**Comment by RedCard1nal** [comment]
I tried enabling vertical sync and it is still flickering. Also I was under the impression that drawing the raw frames in the preview used CPU not GPU. Is that wrong?

**Comment by clsid2** [comment]
Vertical Sync only has effect on the display of frames by the renderer. It is completely unrelated to (hardware accelerated) decoding.

**Comment by clsid2** [comment]
Perhaps any software installed that can capture the screen? For example Discord. Those can sometimes causes issues. Or anything else that is running in the background that might possibly have any effect?

Just brainstorming ideas.

**Comment by RedCard1nal** [comment]
It seems like it is related to mpc transitions from one preview frame to another. If I carefully move the cursor across the seekbar in controlled increments I can transition from one frame to another without any flicker effect, but if I move it in increments too small it flickers and shows the same frame again, or if I just move it quickly across the seekbar it shows the frames with a lot of flickering.

I'm using the behavior of potplayer previews as my reference point, because it seems that the potplayer preview has a way of only showing the current frame preview until it has fetched another frame and is able to display it. With mpc the preview just seems to black when it can't find a frame to display.

**Comment by adipose** [comment]
My theory may be consistent with what @RedCard1nal is saying.  I'm thinking the keyframe search is failing or something, causing a blip until it finds a good frame to display.  mpc-be used a different keyframe search algorithm, which I brought over for the preview, but maybe something in mpc-hc is not compatible.

**Comment by RedCard1nal** [comment]
Is the seek preview frame related to the "fast seek" setting?

**Comment by adipose** [comment]
I think that is used for seeking with the mouse wheel. But it could affect this, I haven't checked yet.

**Comment by MadokaAyukawa** [comment]
Mine flickers with the "fast seek" setting set to "off".

**Comment by clsid2** [comment]
Preview always uses fast seek, except when you hold down Shift.

**Comment by clsid2** [comment]
Test build with some small changes:
https://www.sendspace.com/file/uof0yu

**Comment by MadokaAyukawa** [comment]
This test build still flickers on the Win 10 machine.

**Comment by clsid2** [comment]
> Mine flickers with the "fast seek" setting set to "off".

So you mean it does not flicker with "fast seek" enabled?

**Comment by clsid2** [comment]
I also wonder if the speed of the harddrive is related to the problem. If any of you also have a SSD, test if the problem occurs when loading file from that drive.

**Comment by adipose** [comment]
> I also wonder if the speed of the harddrive is related to the problem. If any of you also have a SSD, test if the problem occurs when loading file from that drive.

I do not have any spinning disks. So that could definitely be related. I wonder if I can artificially introduce a seek time.

**Comment by clsid2** [comment]
I tried with a m2ts file which has slow seeking, and I still don't get any flickering. The preview just updates slow.

**Comment by MadokaAyukawa** [comment]
> 
> 
> > Mine flickers with the "fast seek" setting set to "off".
> 
> So you mean it does not flicker with "fast seek" enabled?

No, it flickers regardless of that setting.

**Comment by clsid2** [comment]
I am beginning to doubt that the problem is related to keyframes and seek speed. Maybe it is an issue with the video renderer.

**Comment by adipose** [comment]
What if you force VMR?

**Comment by RedCard1nal** [comment]
Tested the same video on a mechanical drive and an SSD. It still flickers.

**Comment by clsid2** [comment]
> What if you force VMR9?

Then I do see some rendering issues:
![vmr_issue1](https://user-images.githubusercontent.com/4704996/104823683-4139d500-584c-11eb-9d14-354b8139723e.png)
I also tried with system default renderer (VMR7). Same issue.

**Comment by adipose** [comment]
I was able to slow it down enough to observe the flicker, by using RFS and introducing a delay in the read.  In debugging, I found that the view goes black whenever BeginFlush is sent (possibly due to how it's picked up by EVR).

![image](https://user-images.githubusercontent.com/3324395/105126848-789ac280-5a94-11eb-86bc-d29c3011981f.png)

It may be that a seek normally erases the bg before drawing, but it's not normally so obvious in the video window.

**Comment by clsid2** [comment]
Interesting. Does this flicker also happen with the main video display if using EVR and delayed read?

I have checked if there are any config option for EVR that might affect this. But I only see an option for repaint in stopped state.
https://docs.microsoft.com/en-us/windows/win32/api/evr/ne-evr-mfvideorenderprefs

There is something that is on my wish list for investigating, and that is: what exactly happens when we trigger multiple seeks in quick succession? My idea is to perform a seek. Then wait until it is fully done (if that is possible to track). If any new seeks are triggered in the meantime, then only store that new position value (overwriting existing queued value) and return. Then once the active seek is complete, perform the last queued seek. That should hopefully improve performance and stability. Because when people have crash of freeze issues it is often due to quick seeking.

**Comment by adipose** [comment]
> 
> 
> Interesting. Does this flicker also happen with the main video display if using EVR and delayed read?
> 
> I have checked if there are any config option for EVR that might affect this. But I only see an option for repaint in stopped state.
> https://docs.microsoft.com/en-us/windows/win32/api/evr/ne-evr-mfvideorenderprefs
> 
> There is something that is on my wish list for investigating, and that is: what exactly happens when we trigger multiple seeks in quick succession? My idea is to perform a seek. Then wait until it is fully done (if that is possible to track). If any new seeks are triggered in the meantime, then only store that new position value (overwriting existing queued value) and return. Then once the active seek is complete, perform the last queued seek. That should hopefully improve performance and stability. Because when people have crash of freeze issues it is often due to quick seeking.

I can't get it to go black on EVR for the main video.  Maybe something to do with the preview window moving at the same time.

**Comment by adipose** [comment]
https://github.com/clsid2/mpc-hc/pull/855

**Comment by MadokaAyukawa** [comment]
Test build 148 no longer flickers on my Win 10 machine.

**Comment by RedCard1nal** [comment]
+1 There is also no flickering for me on test build 148. Awesome! A couple of observations:
1. The default behavior of fast seek (nearest keyframe) results in a discrepancy (maybe 6-8 secs) between the frame displayed in preview and the frame it seeks to. The frame it seeks to is further in the video than the preview frame, and the timecode displayed in the preview is about 5 seconds ahead.
2. With fask seek (latest keyframe) the player seems to perfectly seek to the frame displayed in the preview, and the timecode displayed is the preview is several seconds behind the actual time it seeks to.
3. Disabling fast seek just seems to make the whole process less accurate and more variable.

So basically one can choose between the two fast seek setting depending on whether they refer more to timecodes or frame previews for their seeking. Maybe all of this is redundant though because anyone using the preview feature probably refers to frames.

**Comment by adipose** [comment]
https://github.com/clsid2/mpc-hc/pull/866

This has some fixes for VMR9.  Also a compile error due to RFS being required to compile in Lite mode, now.

**Comment by ale5000-git** [comment]
@adipose

I'm testing MPC-HC 1.9.9.2 on Windows 7 now.
Why when the seek preview isn't ready it will show the video under it (transparent preview window) instead of showing the previously previewed image?

See: 
![seek-preview](https://user-images.githubusercontent.com/15793015/107755694-74746480-6d23-11eb-8c32-c554f0dacba9.png)

**Comment by clsid2** [comment]
Wrong keyframe issue has been fixed.

**Comment by ale5000-git** [comment]
Although there is no more flicker when you move from one place with a preview to another place with a preview there is still the problem when you move away so the preview window is hidden and then move back to the preview area, it then remain transparent for a while.

**Comment by narcolepticinsomniac** [comment]
> in addition I also managed to get the preview window transparent

I've seen this fairly often as well. Difficult to provide exact steps to repro, but anecdotally, it seems to always occur when I've been watching in fullscreen without moving the mouse at all for an extended period, and then move directly over the timeline. There's a noticeable glitch where playback stutters, and then all timeline thumbs are transparent/missing.

Also might be worth mentioning, moving strictly vertically up/down on/off the timeline, with no horizontal movement at all, causes the tooltip to fail to appear at all. This is only the case when previews are enabled.

**Comment by adipose** [comment]
> 
> 
> > in addition I also managed to get the preview window transparent
> 
> I've seen this fairly often as well. Difficult to provide exact steps to repro, but anecdotally, it seems to always occur when I've been watching in fullscreen without moving the mouse at all for an extended period, and then move directly over the timeline. There's a noticeable glitch where playback stutters, and then all timeline thumbs are transparent/missing.
> 
> Also might be worth mentioning, moving strictly vertically up/down on/off the timeline, with no horizontal movement at all, causes the tooltip to fail to appear at all. This is only the case when previews are enabled.

https://github.com/clsid2/mpc-hc/pull/1350

This should fix the identical x position problem.

**Comment by ale5000-git** [comment]
@adipose
Isn't possible to detect if there is already a previous preview image and if there isn't then draw a black box until is ready?

To avoid this:
![transparent-bg](https://user-images.githubusercontent.com/15793015/137503179-716ccc77-7d6f-4170-adaf-8d745ad7fd84.png)

**Comment by adipose** [comment]
Painting is done on demand so it might not be so easy.

**Comment by ale5000-git** [comment]
Instead drawing a black box simply behind the preview? (similar to the code of the playlist)

**Comment by adipose** [comment]
Look at CPreview. It fills the bg.

**Comment by adipose** [comment]
I left a video window open for a week and came back to find the preview didn't show (transparent box).  I don't know yet what the reason could be.

**Comment by clsid2** [comment]
We maybe need to also  use `m_pCAP2_preview` in ResetDevice() and DisplayChange()? If your system slept/hibernated during that time.

Another way to reproduce transparent box is by enabling preview for online streams, and then seeking in a Youtube video. Then it happens due to slowness in seeking/decoding.

**Comment by adipose** [comment]
#1359

I did some research and found that the video window area is blanked out by a raw CWND which does no painting. This is a patch to paint that subwindow black. In testing, it seems to work ok without flicker, etc. This should be a solution to the empty window when it can't load frames.

At least in debug, the paint could bleed through on the last line of pixels of the video (when testing with a bright color). It seems to suggest the renderer failed to draw a line, leaving the background showing. I'm not sure how to solve this, but black is as good as transparent, I suppose, in that event.

**Comment by adipose** [comment]
It seems to have a bit of flicker.  Perhaps the "delayed output rect" from the main childview is a better approach.

**Comment by adipose** [comment]
I updated the patch to only paint one time, right after showing the window.  This eliminates any flicker while seeking.  So far, it looks ok.

Maybe if we do clsid2's idea, it won't be necessary, but at least there's a fill, now.

**Comment by adipose** [comment]
> 
> 
> We maybe need to also use `m_pCAP2_preview` in ResetDevice() and DisplayChange()? If your system slept/hibernated during that time.
> 
> Another way to reproduce transparent box is by enabling preview for online streams, and then seeking in a Youtube video. Then it happens due to slowness in seeking/decoding.

Allowing my computer to sleep did cause the problem.  However, I don't actually set my computer to sleep or hibernate regularly, so I don't know exactly what caused it the first time.

**Comment by adipose** [comment]
Latest commit to the patch seems to have fixed the sleep problem, at least, now.
